src/mango/src/mango_indexer.erl

```diff
@@ -0,0 +1,130 @@
+% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+% License for the specific language governing permissions and limitations under
+% the License.
+
+
+-module(mango_indexer).
+
+
+-export([
+    modify/5
+]).
+
+
+-include_lib("couch/include/couch_db.hrl").
+-include("mango.hrl").
+-include("mango_idx.hrl").
+
+-include_lib("couch_mrview/include/couch_mrview.hrl").
+-include_lib("couch_views/include/couch_views.hrl").
+
+
+modify(Db, Change, Doc, PrevDoc, Seq) ->
+    try
+        Indexes = json_indexes(Db, Doc),
+        modify_int(Db, Change, Doc, Indexes, Seq)
+    catch
+        Error:Reason ->
+            DbName = fabric2_db:name(Db),
+            io:format("ERROR ~p ~p ~p ~n", [Error, Reason, erlang:display(erlang:get_stacktrace())]),
+            Id = doc_id(Doc, PrevDoc),
+            couch_log:error("Mango index error for Db ~s Doc ~p ~p ~p",
+                [DbName, Id, Error, Reason])
+    end.
+
+
+doc_id(undefined, #doc{id = DocId}) ->
+    DocId;
+doc_id(undefined, _) ->
+    <<"unknown_doc_id">>;
+doc_id(#doc{id = DocId}, _) ->
+    DocId.
+
+
+% Check if design doc is mango index and kick off background worker
+% to build the new index
+modify_int(Db, Change, #doc{id = <<?DESIGN_DOC_PREFIX, _/binary>>} = Doc,
+        Indexes, Seq) when Change == created orelse Change == updated ->
+    DbName = fabric2_db:name(Db),
+
+    {Props} = couch_doc:to_json_obj(Doc, []),
+    case proplists:get_value(<<"language">>, Props) of
+        <<"query">> ->
+            {ok, Mrst} = couch_mrview_util:ddoc_to_mrst(DbName, Doc),
+            couch_views_fdb:create_build_vs(Db, Mrst),
+            {ok, _} = couch_views_jobs:build_view_async(Db, Mrst, true);
+        _ ->
+            ok
+    end,
+    update_indexes_seq(Db, Indexes, Seq);
+
+modify_int(Db, _Change, Doc, Indexes, Seq) ->
+    write_doc(Db, Doc, Indexes, Seq).
+
+
+write_doc(Db, #doc{deleted = Deleted} = Doc, Indexes, Seq) ->
+    #doc{id = DocId} = Doc,
+    JsonDoc = mango_json:to_binary(couch_doc:to_json_obj(Doc, [])),
+
+    lists:foreach(fun (Idx) ->
+        DocResult0 = #{
+            id => DocId,
+            results => []
+        },
+
+        DocResult1 = case Deleted of
+            true ->
+                DocResult0#{deleted => true};
+            false ->
+                Results = mango_eval:index_doc([Idx], JsonDoc),
+                DocResult0#{results => Results}
+        end,
+
+        DbName = mango_idx:dbname(Idx),
+        DDoc = mango_idx:ddoc(Idx),
+        {ok, Mrst} = couch_mrview_util:ddoc_to_mrst(DbName, DDoc),
+        couch_views_indexer:write_docs(Db, Mrst, [DocResult1], #{last_seq => Seq})
+    end, Indexes).
+
+json_indexes(Db,  #doc{id = <<?DESIGN_DOC_PREFIX, _/binary>> = Id}) ->
+    json_indexes(Db, [{ignore, [Id]}]);
+
+json_indexes(Db,  #doc{}) ->
+    json_indexes(Db, []);
+
+json_indexes(Db, Opts) when is_list(Opts) ->
+    #{
+        md_version := MdVersion
+    } = Db,
+
+    case get(json_indexes) of
```

Comment:
Pretty sure this is a bit of a nasty bug. If a client were to do multiple request across HTTP/1.1 then this could be cached across transactions and refer to the wrong database. Removing it between transactions also means that this is only caching for multiple updates in a single interactive _bulk_docs write. And thirdly, they'd already be cached on the transaction object so its not saving that much overhead since it'd just be the transformation to #mrst{} records rather than round trips to the database.

Reply:
Good catch, I didn't give this enough thought. I've added in the dbname as another part of the tuple. When I run the mango test `20-no-timeout-test.py` it will fail to add 1000 docs in a transaction unless I cache the indexes here. It seems there is quite a lot of overhead for continuously reading the same keys. Locally on my machine I see it taking about 400ms per 1000 docs per transaction and that goes to over 1s if I don't cache the indexes.

Reply:
That test is likely a not very good test for this caching. Theoretically we should cache those indexes on the db handle so we're only re-reading when a design doc changes instead of on every new HTTP connection.

Reply:
Errr, to be specific, its not an awesome test cause its probably re-using an HTTP/1.1 connection for each _bulk_docs request whcih means its hiding that the caching doesn't have an effect between multiple clients.